git url parsing unit test failure with golang-1.8

When using golang 1.8, there are failures related to parsing git urls due to changes in url.Parse in golang 1.8

##### Version
master

##### Steps To Reproduce
1. OS_BUILD_ENV_GOLANG=1.8 hack/env make release
2. OS_BUILD_ENV_GOLANG=1.8 hack/env make test-unit


##### Current Result
```
--- FAIL: TestParseRepository (0.08s)
	git_test.go:101: ParseRepository returned err: parse git@[2001:db8::1]:repository.git: first path segment in URL cannot contain colon
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x229c48]

goroutine 7 [running]:
testing.tRunner.func1(0xc42007d110)
	/usr/lib/golang/src/testing/testing.go:622 +0x25c
panic(0x274e40, 0x429700)
	/usr/lib/golang/src/runtime/panic.go:489 +0x284
github.com/openshift/origin/pkg/generate/git.TestParseRepository(0xc42007d110)
	/root/origin/_output/local/go/src/github.com/openshift/origin/pkg/generate/git/git_test.go:108 +0xf58
testing.tRunner(0xc42007d110, 0x2d7278)
	/usr/lib/golang/src/testing/testing.go:657 +0x9c
created by testing.(*T).Run
	/usr/lib/golang/src/testing/testing.go:697 +0x270
FAIL	github.com/openshift/origin/pkg/generate/git	0.137s
```

##### Expected Result
Unit test cases in pkg/generate/git/git_test.go should pass.

--- a/Godeps/Godeps.json
+++ b/Godeps/Godeps.json
@@ -2967,118 +2967,138 @@
 		},
 		{
 			"ImportPath": "github.com/openshift/source-to-image/pkg/api",
-			"Comment": "v1.1.5-31-g191ae3b",
-			"Rev": "7f58756254b0a65bf59fa87a8ecedad01ce6a85b"
+			"Comment": "v1.1.7-20-g7da3d3e",
+			"Rev": "7da3d3e97565e652a59fd81286f3956fd43e85a8"
 		},
 		{
 			"ImportPath": "github.com/openshift/source-to-image/pkg/api/describe",
-			"Comment": "v1.1.5-31-g191ae3b",
-			"Rev": "7f58756254b0a65bf59fa87a8ecedad01ce6a85b"
+			"Comment": "v1.1.7-20-g7da3d3e",
+			"Rev": "7da3d3e97565e652a59fd81286f3956fd43e85a8"
 		},
 		{
 			"ImportPath": "github.com/openshift/source-to-image/pkg/api/validation",
-			"Comment": "v1.1.5-31-g191ae3b",
-			"Rev": "7f58756254b0a65bf59fa87a8ecedad01ce6a85b"
+			"Comment": "v1.1.7-20-g7da3d3e",
+			"Rev": "7da3d3e97565e652a59fd81286f3956fd43e85a8"
 		},
 		{
 			"ImportPath": "github.com/openshift/source-to-image/pkg/build",
-			"Comment": "v1.1.5-31-g191ae3b",
-			"Rev": "7f58756254b0a65bf59fa87a8ecedad01ce6a85b"
+			"Comment": "v1.1.7-20-g7da3d3e",
+			"Rev": "7da3d3e97565e652a59fd81286f3956fd43e85a8"
 		},
 		{
 			"ImportPath": "github.com/openshift/source-to-image/pkg/build/strategies",
-			"Comment": "v1.1.5-31-g191ae3b",
-			"Rev": "7f58756254b0a65bf59fa87a8ecedad01ce6a85b"
+			"Comment": "v1.1.7-20-g7da3d3e",
+			"Rev": "7da3d3e97565e652a59fd81286f3956fd43e85a8"
 		},
 		{
 			"ImportPath": "github.com/openshift/source-to-image/pkg/build/strategies/layered",
-			"Comment": "v1.1.5-31-g191ae3b",
-			"Rev": "7f58756254b0a65bf59fa87a8ecedad01ce6a85b"
+			"Comment": "v1.1.7-20-g7da3d3e",
+			"Rev": "7da3d3e97565e652a59fd81286f3956fd43e85a8"
 		},
 		{
 			"ImportPath": "github.com/openshift/source-to-image/pkg/build/strategies/onbuild",
-			"Comment": "v1.1.5-31-g191ae3b",
-			"Rev": "7f58756254b0a65bf59fa87a8ecedad01ce6a85b"
+			"Comment": "v1.1.7-20-g7da3d3e",
+			"Rev": "7da3d3e97565e652a59fd81286f3956fd43e85a8"
 		},
 		{
 			"ImportPath": "github.com/openshift/source-to-image/pkg/build/strategies/sti",
-			"Comment": "v1.1.5-31-g191ae3b",
-			"Rev": "7f58756254b0a65bf59fa87a8ecedad01ce6a85b"
+			"Comment": "v1.1.7-20-g7da3d3e",
+			"Rev": "7da3d3e97565e652a59fd81286f3956fd43e85a8"
 		},
 		{
 			"ImportPath": "github.com/openshift/source-to-image/pkg/docker",
-			"Comment": "v1.1.5-31-g191ae3b",
-			"Rev": "7f58756254b0a65bf59fa87a8ecedad01ce6a85b"
+			"Comment": "v1.1.7-20-g7da3d3e",
+			"Rev": "7da3d3e97565e652a59fd81286f3956fd43e85a8"
 		},
 		{
 			"ImportPath": "github.com/openshift/source-to-image/pkg/errors",
-			"Comment": "v1.1.5-31-g191ae3b",
-			"Rev": "7f58756254b0a65bf59fa87a8ecedad01ce6a85b"
+			"Comment": "v1.1.7-20-g7da3d3e",
+			"Rev": "7da3d3e97565e652a59fd81286f3956fd43e85a8"
 		},
 		{
 			"ImportPath": "github.com/openshift/source-to-image/pkg/ignore",
-			"Comment": "v1.1.5-31-g191ae3b",
-			"Rev": "7f58756254b0a65bf59fa87a8ecedad01ce6a85b"
+			"Comment": "v1.1.7-20-g7da3d3e",
+			"Rev": "7da3d3e97565e652a59fd81286f3956fd43e85a8"
 		},
 		{
 			"ImportPath": "github.com/openshift/source-to-image/pkg/scm",
-			"Comment": "v1.1.5-31-g191ae3b",
-			"Rev": "7f58756254b0a65bf59fa87a8ecedad01ce6a85b"
+			"Comment": "v1.1.7-20-g7da3d3e",
+			"Rev": "7da3d3e97565e652a59fd81286f3956fd43e85a8"
 		},
 		{
-			"ImportPath": "github.com/openshift/source-to-image/pkg/scm/empty",
-			"Comment": "v1.1.5-31-g191ae3b",
-			"Rev": "7f58756254b0a65bf59fa87a8ecedad01ce6a85b"
+			"ImportPath": "github.com/openshift/source-to-image/pkg/scm/downloaders",
+			"Comment": "v1.1.7-20-g7da3d3e",
+			"Rev": "7da3d3e97565e652a59fd81286f3956fd43e85a8"
 		},
 		{
-			"ImportPath": "github.com/openshift/source-to-image/pkg/scm/file",
-			"Comment": "v1.1.5-31-g191ae3b",
-			"Rev": "7f58756254b0a65bf59fa87a8ecedad01ce6a85b"
+			"ImportPath": "github.com/openshift/source-to-image/pkg/scm/downloaders/empty",
+			"Comment": "v1.1.7-20-g7da3d3e",
+			"Rev": "7da3d3e97565e652a59fd81286f3956fd43e85a8"
+		},
+		{
+			"ImportPath": "github.com/openshift/source-to-image/pkg/scm/downloaders/file",
+			"Comment": "v1.1.7-20-g7da3d3e",
+			"Rev": "7da3d3e97565e652a59fd81286f3956fd43e85a8"
+		},
+		{
+			"ImportPath": "github.com/openshift/source-to-image/pkg/scm/downloaders/git",
+			"Comment": "v1.1.7-20-g7da3d3e",
+			"Rev": "7da3d3e97565e652a59fd81286f3956fd43e85a8"
 		},
 		{
 			"ImportPath": "github.com/openshift/source-to-image/pkg/scm/git",
-			"Comment": "v1.1.5-31-g191ae3b",
-			"Rev": "7f58756254b0a65bf59fa87a8ecedad01ce6a85b"
+			"Comment": "v1.1.7-20-g7da3d3e",
+			"Rev": "7da3d3e97565e652a59fd81286f3956fd43e85a8"
 		},
 		{
 			"ImportPath": "github.com/openshift/source-to-image/pkg/scripts",
-			"Comment": "v1.1.5-31-g191ae3b",
-			"Rev": "7f58756254b0a65bf59fa87a8ecedad01ce6a85b"
+			"Comment": "v1.1.7-20-g7da3d3e",
+			"Rev": "7da3d3e97565e652a59fd81286f3956fd43e85a8"
 		},
 		{
 			"ImportPath": "github.com/openshift/source-to-image/pkg/tar",
-			"Comment": "v1.1.5-31-g191ae3b",
-			"Rev": "7f58756254b0a65bf59fa87a8ecedad01ce6a85b"
-		},
-		{
-			"ImportPath": "github.com/openshift/source-to-image/pkg/test",
-			"Comment": "v1.1.5-31-g191ae3b",
-			"Rev": "7f58756254b0a65bf59fa87a8ecedad01ce6a85b"
+			"Comment": "v1.1.7-20-g7da3d3e",
+			"Rev": "7da3d3e97565e652a59fd81286f3956fd43e85a8"
 		},
 		{
 			"ImportPath": "github.com/openshift/source-to-image/pkg/util",
-			"Comment": "v1.1.5-31-g191ae3b",
-			"Rev": "7f58756254b0a65bf59fa87a8ecedad01ce6a85b"
+			"Comment": "v1.1.7-20-g7da3d3e",
+			"Rev": "7da3d3e97565e652a59fd81286f3956fd43e85a8"
+		},
+		{
+			"ImportPath": "github.com/openshift/source-to-image/pkg/util/cmd",
+			"Comment": "v1.1.7-20-g7da3d3e",
+			"Rev": "7da3d3e97565e652a59fd81286f3956fd43e85a8"
+		},
+		{
+			"ImportPath": "github.com/openshift/source-to-image/pkg/util/cygpath",
+			"Comment": "v1.1.7-20-g7da3d3e",
+			"Rev": "7da3d3e97565e652a59fd81286f3956fd43e85a8"
+		},
+		{
+			"ImportPath": "github.com/openshift/source-to-image/pkg/util/fs",
+			"Comment": "v1.1.7-20-g7da3d3e",
+			"Rev": "7da3d3e97565e652a59fd81286f3956fd43e85a8"
 		},
 		{
 			"ImportPath": "github.com/openshift/source-to-image/pkg/util/glog",
-			"Comment": "v1.1.5-31-g191ae3b",
-			"Rev": "7f58756254b0a65bf59fa87a8ecedad01ce6a85b"
+			"Comment": "v1.1.7-20-g7da3d3e",
+			"Rev": "7da3d3e97565e652a59fd81286f3956fd43e85a8"
 		},
 		{
 			"ImportPath": "github.com/openshift/source-to-image/pkg/util/interrupt",
-			"Comment": "v1.1.5-31-g191ae3b",
-			"Rev": "7f58756254b0a65bf59fa87a8ecedad01ce6a85b"
+			"Comment": "v1.1.7-20-g7da3d3e",
+			"Rev": "7da3d3e97565e652a59fd81286f3956fd43e85a8"
 		},
 		{
 			"ImportPath": "github.com/openshift/source-to-image/pkg/util/status",
-			"Comment": "v1.1.5-31-g191ae3b",
-			"Rev": "7f58756254b0a65bf59fa87a8ecedad01ce6a85b"
+			"Comment": "v1.1.7-20-g7da3d3e",
+			"Rev": "7da3d3e97565e652a59fd81286f3956fd43e85a8"
 		},
 		{
 			"ImportPath": "github.com/openshift/source-to-image/pkg/util/user",
-			"Comment": "v1.1.5-31-g191ae3b",
-			"Rev": "7f58756254b0a65bf59fa87a8ecedad01ce6a85b"
+			"Comment": "v1.1.7-20-g7da3d3e",
+			"Rev": "7da3d3e97565e652a59fd81286f3956fd43e85a8"
 		},
 		{
 			"ImportPath": "github.com/pborman/uuid",

--- a/pkg/build/apis/build/validation/validation.go
+++ b/pkg/build/apis/build/validation/validation.go
@@ -2,7 +2,6 @@ package validation
 
 import (
 	"fmt"
-	"net/url"
 	"path"
 	"path/filepath"
 	"strings"
@@ -24,6 +23,7 @@ import (
 	imageapi "github.com/openshift/origin/pkg/image/apis/image"
 	imageapivalidation "github.com/openshift/origin/pkg/image/apis/image/validation"
 	"github.com/openshift/origin/pkg/util/labelselector"
+	s2igit "github.com/openshift/source-to-image/pkg/scm/git"
 )
 
 // ValidateBuild tests required fields for a Build.
@@ -271,14 +271,18 @@ func validateGitSource(git *buildapi.GitBuildSource, fldPath *field.Path) field.
 	allErrs := field.ErrorList{}
 	if len(git.URI) == 0 {
 		allErrs = append(allErrs, field.Required(fldPath.Child("uri"), ""))
-	} else if !IsValidURL(git.URI) {
-		allErrs = append(allErrs, field.Invalid(fldPath.Child("uri"), git.URI, "uri is not a valid url"))
+	} else if _, err := s2igit.Parse(git.URI); err != nil {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("uri"), git.URI, err.Error()))
 	}
-	if git.HTTPProxy != nil && len(*git.HTTPProxy) != 0 && !IsValidURL(*git.HTTPProxy) {
-		allErrs = append(allErrs, field.Invalid(fldPath.Child("httpproxy"), *git.HTTPProxy, "proxy is not a valid url"))
+	if git.HTTPProxy != nil && len(*git.HTTPProxy) != 0 {
+		if _, err := buildutil.ParseProxyURL(*git.HTTPProxy); err != nil {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("httpproxy"), *git.HTTPProxy, err.Error()))
+		}
 	}
-	if git.HTTPSProxy != nil && len(*git.HTTPSProxy) != 0 && !IsValidURL(*git.HTTPSProxy) {
-		allErrs = append(allErrs, field.Invalid(fldPath.Child("httpsproxy"), *git.HTTPSProxy, "proxy is not a valid url"))
+	if git.HTTPSProxy != nil && len(*git.HTTPSProxy) != 0 {
+		if _, err := buildutil.ParseProxyURL(*git.HTTPSProxy); err != nil {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("httpsproxy"), *git.HTTPSProxy, err.Error()))
+		}
 	}
 	return allErrs
 }
@@ -635,11 +639,6 @@ func validateWebHook(webHook *buildapi.WebHookTrigger, fldPath *field.Path, isGe
 		allErrs = append(allErrs, field.Invalid(fldPath.Child("allowEnv"), webHook, "git webhooks cannot allow env vars"))
 	}
 	return allErrs
-}
-
-func IsValidURL(uri string) bool {
-	_, err := url.Parse(uri)
-	return err == nil
 }
 
 func ValidateBuildLogOptions(opts *buildapi.BuildLogOptions) field.ErrorList {

--- a/pkg/build/apis/build/validation/validation_test.go
+++ b/pkg/build/apis/build/validation/validation_test.go
@@ -841,7 +841,7 @@ func TestValidateSource(t *testing.T) {
 			path: "git.uri",
 			source: &buildapi.BuildSource{
 				Git: &buildapi.GitBuildSource{
-					URI: "::",
+					URI: "http://%",
 				},
 			},
 		},

--- a/pkg/build/builder/sti.go
+++ b/pkg/build/builder/sti.go
@@ -440,14 +440,14 @@ func scriptProxyConfig(build *buildapi.Build) (*s2iapi.ProxyConfig, error) {
 	}
 	config := &s2iapi.ProxyConfig{}
 	if len(httpProxy) > 0 {
-		proxyURL, err := url.Parse(httpProxy)
+		proxyURL, err := util.ParseProxyURL(httpProxy)
 		if err != nil {
 			return nil, err
 		}
 		config.HTTPProxy = proxyURL
 	}
 	if len(httpsProxy) > 0 {
-		proxyURL, err := url.Parse(httpsProxy)
+		proxyURL, err := util.ParseProxyURL(httpsProxy)
 		if err != nil {
 			return nil, err
 		}

--- a/pkg/build/controller/build/defaults/api/validation/validation.go
+++ b/pkg/build/controller/build/defaults/api/validation/validation.go
@@ -6,13 +6,14 @@ import (
 
 	buildvalidation "github.com/openshift/origin/pkg/build/apis/build/validation"
 	"github.com/openshift/origin/pkg/build/controller/build/defaults/api"
+	"github.com/openshift/origin/pkg/build/util"
 )
 
 // ValidateBuildDefaultsConfig tests required fields for a Build.
 func ValidateBuildDefaultsConfig(config *api.BuildDefaultsConfig) field.ErrorList {
 	allErrs := field.ErrorList{}
-	allErrs = append(allErrs, validateURL(config.GitHTTPProxy, field.NewPath("gitHTTPProxy"))...)
-	allErrs = append(allErrs, validateURL(config.GitHTTPSProxy, field.NewPath("gitHTTPSProxy"))...)
+	allErrs = append(allErrs, validateProxyURL(config.GitHTTPProxy, field.NewPath("gitHTTPProxy"))...)
+	allErrs = append(allErrs, validateProxyURL(config.GitHTTPSProxy, field.NewPath("gitHTTPSProxy"))...)
 	allErrs = append(allErrs, buildvalidation.ValidateStrategyEnv(config.Env, field.NewPath("env"))...)
 	allErrs = append(allErrs, buildvalidation.ValidateImageLabels(config.ImageLabels, field.NewPath("imageLabels"))...)
 	allErrs = append(allErrs, buildvalidation.ValidateNodeSelector(config.NodeSelector, field.NewPath("nodeSelector"))...)
@@ -21,10 +22,10 @@ func ValidateBuildDefaultsConfig(config *api.BuildDefaultsConfig) field.ErrorLis
 }
 
 //
-func validateURL(u string, path *field.Path) field.ErrorList {
+func validateProxyURL(u string, path *field.Path) field.ErrorList {
 	allErrs := field.ErrorList{}
-	if !buildvalidation.IsValidURL(u) {
-		allErrs = append(allErrs, field.Invalid(path, u, "invalid URL"))
+	if _, err := util.ParseProxyURL(u); err != nil {
+		allErrs = append(allErrs, field.Invalid(path, u, err.Error()))
 	}
 	return allErrs
 }

--- a/pkg/build/util/util.go
+++ b/pkg/build/util/util.go
@@ -461,3 +461,23 @@ func FindDockerSecretAsReference(secrets []kapi.Secret, image string) *kapi.Loca
 	}
 	return nil
 }
+
+// ParseProxyURL parses a proxy URL and allows fallback to non-URLs like
+// myproxy:80 (for example) which url.Parse no longer accepts in Go 1.8.  The
+// logic is copied from net/http.ProxyFromEnvironment to try to maintain
+// backwards compatibility.
+func ParseProxyURL(proxy string) (*url.URL, error) {
+	proxyURL, err := url.Parse(proxy)
+
+	// logic copied from net/http.ProxyFromEnvironment
+	if err != nil || !strings.HasPrefix(proxyURL.Scheme, "http") {
+		// proxy was bogus. Try prepending "http://" to it and see if that
+		// parses correctly. If not, we fall through and complain about the
+		// original one.
+		if proxyURL, err := url.Parse("http://" + proxy); err == nil {
+			return proxyURL, nil
+		}
+	}
+
+	return proxyURL, err
+}

--- a/test/extended/testdata/bindata.go
+++ b/test/extended/testdata/bindata.go
@@ -7438,7 +7438,7 @@ spec:
       scripts: "http://example.org/scripts"
       env:
         - name: http_proxy
-          value: ":http://example.org"
+          value: "http://%"
 `)
 
 func testExtendedTestdataStatusfailGenericreasonYamlBytes() ([]byte, error) {

--- a/test/extended/testdata/statusfail-genericreason.yaml
+++ b/test/extended/testdata/statusfail-genericreason.yaml
@@ -15,4 +15,4 @@ spec:
       scripts: "http://example.org/scripts"
       env:
         - name: http_proxy
-          value: ":http://example.org"
+          value: "http://%"

